### PR TITLE
feat(raiko): put the tasks that cannot run in parallel into pending list

### PR DESCRIPTION
**Pull request by smtmfft**
image

### CODE REVIEW

File: host/src/proof.rs [hunk 1]
```diff
@@ -122,25 +139,56 @@ impl ProofActor {
...
+                        );
+                        let mut pending_tasks = self.pending_tasks.lock().await;
+                        pending_tasks.push_back(proof_request);
+                    }
+                }
+                Message::TaskComplete(req) => {
+                    // pop up pending task if any task complete
+                    debug!("Message::TaskComplete: {:?}", req);
+                    info!(
+                        "task completed, current runing {:?}, pending: {:?}",
```
Comment (petarvujovic98) [inline host/src/proof.rs:180]:
```suggestion
                        "task completed, current running {:?}, pending: {:?}",
```

File: host/src/proof.rs [hunk 2]
```diff
@@ -122,25 +139,56 @@ impl ProofActor {
...
-                        .expect("Couldn't acquire permit");
-                    self.run_task(proof_request, permit).await;
+                    debug!("Message::Task proof_request: {:?}", proof_request);
+                    let running_task_count = self.running_tasks.lock().await.len();
+                    if running_task_count < self.opts.concurrency_limit {
+                        info!("Running task {:?}", proof_request);
+                        self.run_task(proof_request).await;
+                    } else {
+                        info!(
+                            "Task concurrency limit reached, current runing {:?}, pending: {:?}",
```
Comment (petarvujovic98) [inline host/src/proof.rs:168]:
```suggestion
                            "Task concurrency limit reached, current running {:?}, pending: {:?}",
```

**Review by petarvujovic98 — CHANGES_REQUESTED** [review]

File: host/src/lib.rs [hunk 3]
```diff
@@ -146,6 +146,7 @@ pub struct ProverState {
 pub enum Message {
     Cancel(TaskDescriptor),
     Task(ProofRequest),
+    TaskComplete(ProofRequest),
```
[the remark this answers is not in the document]
Comment (smtmfft) [reply host/src/lib.rs:149]: you mean notify one-shot channel & let it send task in??

**Review by johntaiko — APPROVED** [review]

File: host/src/proof.rs [hunk 4]
```diff
@@ -203,21 +226,48 @@ impl ProofActor {
...
+                            running_task_count,
+                            self.pending_tasks.lock().await.len()
+                        );
+                        let mut pending_tasks = self.pending_tasks.lock().await;
+                        pending_tasks.push_back(proof_request);
+                    }
+                }
+                Message::TaskComplete(req) => {
+                    // pop up pending task if any task complete
+                    debug!("Message::TaskComplete: {:?}", req);
```
Comment (petarvujovic98) [inline host/src/proof.rs:257]:
```suggestion
                    debug!("Message::TaskComplete: {req:?}");
```

File: host/src/proof.rs [hunk 5]
```diff
@@ -203,21 +226,48 @@ impl ProofActor {
...
-                    self.run_task(proof_request, permit).await;
+                    debug!("Message::Task proof_request: {:?}", proof_request);
+                    let running_task_count = self.running_tasks.lock().await.len();
+                    if running_task_count < self.opts.concurrency_limit {
+                        info!("Running task {:?}", proof_request);
+                        self.run_task(proof_request).await;
+                    } else {
+                        info!(
+                            "Task concurrency limit reached, current running {:?}, pending: {:?}",
+                            running_task_count,
```
Comment (petarvujovic98) [inline host/src/proof.rs:248]:
```suggestion
                            "Task concurrency limit reached, current running {running_task_count:?}, pending: {:?}",
```

File: host/src/proof.rs [hunk 6]
```diff
@@ -203,21 +226,48 @@ impl ProofActor {
...
                 Message::Task(proof_request) => {
-                    let permit = Arc::clone(&semaphore)
-                        .acquire_owned()
-                        .await
-                        .expect("Couldn't acquire permit");
-                    self.run_task(proof_request, permit).await;
+                    debug!("Message::Task proof_request: {:?}", proof_request);
+                    let running_task_count = self.running_tasks.lock().await.len();
+                    if running_task_count < self.opts.concurrency_limit {
+                        info!("Running task {:?}", proof_request);
```
Comment (petarvujovic98) [inline host/src/proof.rs:243]:
```suggestion
                        info!("Running task {proof_request:?}");
```

File: host/src/proof.rs [hunk 7]
```diff
@@ -203,21 +226,48 @@ impl ProofActor {
...
                         error!("Failed to cancel task: {error}")
                     }
                 }
                 Message::Task(proof_request) => {
-                    let permit = Arc::clone(&semaphore)
-                        .acquire_owned()
-                        .await
-                        .expect("Couldn't acquire permit");
-                    self.run_task(proof_request, permit).await;
+                    debug!("Message::Task proof_request: {:?}", proof_request);
```
Comment (petarvujovic98) [inline host/src/proof.rs:240]:
```suggestion
                    debug!("Message::Task proof_request: {proof_request:?}");
```

File: host/src/proof.rs [hunk 8]
```diff
@@ -203,21 +226,48 @@ impl ProofActor {
     }
 
     pub async fn run(&mut self) {
+        // recv() is protected by outside mpsc, no lock needed here
         let semaphore = Arc::new(Semaphore::new(self.opts.concurrency_limit));
-
         while let Some(message) = self.receiver.recv().await {
             match message {
                 Message::Cancel(key) => {
+                    debug!("Message::Cancel task: {:?}", key);
```
Comment (petarvujovic98) [inline host/src/proof.rs:234]:
```suggestion
                    debug!("Message::Cancel task: {key:?}");
```

**Review by petarvujovic98 — APPROVED** [review]

**Pull request merged**